client/node-common/src/service.rs

```diff
@@ -0,0 +1,652 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Tanssi.  If not, see <http://www.gnu.org/licenses/>.
+
+use {
+    async_io::Timer,
+    core::time::Duration,
+    core_extensions::TypeIdentity,
+    cumulus_client_cli::CollatorOptions,
+    cumulus_client_service::{build_relay_chain_interface, CollatorSybilResistance},
+    cumulus_primitives_core::ParaId,
+    cumulus_relay_chain_interface::RelayChainInterface,
+    frame_benchmarking_cli::SUBSTRATE_REFERENCE_HARDWARE,
+    futures::{channel::mpsc, FutureExt, Stream, StreamExt},
+    jsonrpsee::RpcModule,
+    polkadot_primitives::CollatorPair,
+    sc_client_api::Backend,
+    sc_consensus::{block_import, BlockImport, ImportQueue},
+    sc_consensus_manual_seal::{
+        run_manual_seal, ConsensusDataProvider, EngineCommand, ManualSealParams,
+    },
+    sc_executor::{
+        HeapAllocStrategy, NativeElseWasmExecutor, NativeExecutionDispatch, WasmExecutor,
+        DEFAULT_HEAP_ALLOC_STRATEGY,
+    },
+    sc_network::{config::FullNetworkConfiguration, NetworkService},
+    sc_network_sync::SyncingService,
+    sc_network_transactions::TransactionsHandlerController,
+    sc_rpc::{DenyUnsafe, SubscriptionTaskExecutor},
+    sc_service::{
+        Configuration, KeystoreContainer, NetworkStarter, TFullBackend, TFullClient, TaskManager,
+    },
+    sc_telemetry::{Telemetry, TelemetryWorker, TelemetryWorkerHandle},
+    sc_transaction_pool_api::OffchainTransactionPoolFactory,
+    sc_utils::mpsc::TracingUnboundedSender,
+    sp_api::ConstructRuntimeApi,
+    sp_block_builder::BlockBuilder,
+    sp_consensus::{EnableProofRecording, SelectChain},
+    sp_core::H256,
+    sp_inherents::CreateInherentDataProviders,
+    sp_offchain::OffchainWorkerApi,
+    sp_runtime::Percent,
+    sp_transaction_pool::runtime_api::TaggedTransactionQueue,
+    std::{str::FromStr, sync::Arc},
+};
+
+/// Functions in this module are generic over `Block`, `RuntimeApi`, and
+/// `ParachainNativeExecutor`. Using type aliases requires them to be
+/// generic too, which makes them still verbose to use. For that reason we use
+/// a macro that expect the above types to already be in scope.
+macro_rules! T {
+    [Executor] => { NativeElseWasmExecutor<ParachainNativeExecutor> };
+    [Client] => { TFullClient<Block, RuntimeApi, T![Executor]> };
+    [Backend] => { TFullBackend<Block> };
+    [ConstructedRuntimeApi] => {
+        <RuntimeApi as ConstructRuntimeApi<Block, T![Client]>>::RuntimeApi
+    };
+    [Where] => {
+        Block: cumulus_primitives_core::BlockT,
+        ParachainNativeExecutor: NativeExecutionDispatch + 'static,
+        RuntimeApi: ConstructRuntimeApi<Block, T![Client]> + Sync + Send + 'static,
+        T![ConstructedRuntimeApi]: TaggedTransactionQueue<Block> + BlockBuilder<Block>,
+    }
+}
```

Comment:
This macro looks like a bad idea, I'm mostly worried about the error messages a user will see if they pass some wrong input type (I don't know if the macro will be expanded in the error message or not), as well as the errors we will encounter on the next polkadot upgrade.

Maybe instead we can take a similar approach to the one used in `BlockNumberFor` https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet_prelude/type.BlockNumberFor.html

So instead of making the functions generic over `Block`, `RuntimeApi`, and `ParachainNativeExecutor`, make them generic over `Runtime`, or something else that can be used to derive all the other types. So we would have `ClientFor<R>` instead of `T![Client]`. It may make this module a bit more verbose, but then I believe this crate will be easier to integrate with other projects.

Reply:
In error messages it gives verbose names. The "config trait" approach can be used instead indeed. I'll make the change.